python-avd/pyavd/_cv/workflows/finalize_workspace_on_cv.py

```diff
@@ -87,3 +101,157 @@ async def finalize_workspace_on_cv(workspace: CVWorkspace, cv_client: CVClient)
+    - Config validation errors.
+    - Config validation warnings.
+    - Image validation errors.
+    - Image validation warnings.
+
+    Parameters:
+        workspace: Active Workspace.
+        cv_client: CVClient.
+        devices: List of CVDevice objects representing existing CV devices.
+
+    Returns:
+        List of 'CVWorkspaceBuildResult's describing observed Workspace build validation errors and warnings per targeted device.
+    """
+    try:
+        workspace_build_details_stream_response = await cv_client.get_workspace_build_details(workspace_id=workspace.id, build_id=workspace.build_id)
+
+        if workspace.build_warnings:
+            # Deduplicate list of patterns from user
+            workspace_build_warning_suppress_list = list(set(workspace.build_warnings_suppress_patterns[:]))
+            if workspace.build_warnings_suppress_portfast and EOS_CLI_PORTFAST_WARNING not in workspace_build_warning_suppress_list:
+                workspace_build_warning_suppress_list.append(EOS_CLI_PORTFAST_WARNING)
+            compiled_workspace_build_warning_suppress_list: list = []
+            for proposed_regex_pattern in workspace_build_warning_suppress_list:
+                try:
+                    compiled_workspace_build_warning_suppress_list.append(re_compile(proposed_regex_pattern))
+                except Exception as e:  # noqa: PERF203
+                    LOGGER.warning(
+                        "process_workspace_build_details: Failed to process proposed regex pattern '%s'. "
+                        "This incorrect pattern will not be used for warnings suppression. Error: '%s'",
+                        proposed_regex_pattern,
+                        e,
+                    )
+                    # Delete invalid regex pattern string from workspace.build_warnings_suppress_patterns
+                    workspace.build_warnings_suppress_patterns = [item for item in workspace.build_warnings_suppress_patterns if item != proposed_regex_pattern]
+
+        workspace_build_results = [
+            CVWorkspaceBuildResult(
+                device=next((device for device in devices if device.serial_number == device_id), None)
+                if (device_id := getattr(getattr(getattr(response, "value", None), "key", None), "device_id", None)) is not None
+                else None,
+                stages_states=[
+                    CVWorkspaceBuildStageState(
+                        stage=stage,
+                        state=getattr(state, "name", None),
+                    )
+                    for stage, state in getattr(getattr(getattr(response, "value", None), "build_stage_state", None), "values", {}).items()
+                ],
+                config_validation=CVWorkspaceBuildConfigValidationResult(
+                    errors=[
+                        CVWorkspaceBuildConfigValidationError(
+                            error_code=getattr(getattr(config_validation_error, "error_code", None), "name", None),
```

Comment:
If we really need this, then please use get_v2 under our utils. It shoudl do exactly this.
It would be better to check the type of response early and then trust the API classes generated from the protos.

Reply:
WS build details may just return empty dict in case there are no errors, like

```
values.config_validation_result.errors: {}
```

but if actual errors were faced - it'll return nested dict (with list of ConfigError objects) that we need to loop through

```
values.config_validation_result.errors.values: [
    ConfigError
]
```

Reply:
fixed